2022/incoming/building-the-sleigh-while-flying-it.pod

```diff
@@ -0,0 +1,108 @@
+
+
+The terminal lit up like a Christmas Tree. They could see the full class hierarchy, applied roles, attributes, public and private methods - everything! Santa's eyes sparkled. I<"Ho! Ho! Ho! I know what's wrong."> He pointed at the line that said C<< after  dash_away: <undefined coderef> >> in bright red. They searched the source code for C<< 'after "dash_away"' >> and quickly found the only entry, that read:
+
+	after 'dash_away' => \&lifftoff;
+
+They took another look at the Data::Printer output searching for that method. Santa quickly spotted it coming from the C<FlyingMagic> role, but it was written differently. I<"A typo! Someone was naughty on that code review."> They fixed the typo and ran it again. It worked! One by one the reindeers started whirring. Santa was about to jump on the sleigh for a final test-drive when the elf interrupted the program. I<"What is it, B?"> the old man asked.
+
+Binky was staring at the terminal again. I<"It's leaking memory. The way things are, the sleigh will crash before you even reach the Atlantic. Hold on."> The elf added a bunch of C<p()> statements throughout the code. Santa Claus looked bewildered. I<"What are you doing? There will be so much stuff in the output we won't be able to make any of it out!">, he scoffed.
+
+I<"Sure we will."> Binky replied without taking her eyes off the monitor. I<"We'll control what prints what on the .dataprinter file">. She created a new file called ".dataprinter" on the root of the project path, then made it look something like this:
+
+
+	show_refcount = 1
+	theme = Material
+
+	[Sleigh::Engine]
+	    class.expand = 0
+
+	[Toys::Storage::MagicBag]
+	    filters = DB
+
+	[Reindeer]
+	    class.stringify = 0
+            class.show_wrapped = 0
+            class.inherited = none
+	    output = /var/log/christmas/reindeer.data
+
+Binky went on to explain that you could place global settings for Data::Printer at the top of that file, like setting a theme and showing refcounts, then create more specific rules that apply only to certain caller packages. In this case, any C<p()> statements made inside C<Sleigh::Engine> would not expand objects, showing only the class name and (as per the global setting) the refcount.
+On the other hand, calls to C<p()> inside C<Toys::Storage::MagicBag> would be subject to all the filters set on L<Data::Printer::Filters::DB>, useful to see database connection statuses, prepared statements, result sources and sets from DBI, DBIx::Class and more.
+Finally, calls to C<p()> made inside C<Reindeer.pm> would not attempt to stringify objects that have a subroutine `as_string`, `to_string` or similar, nor will they show wrapped method modifiers or methods inherited from parent classes. Even so, that particular output is in a hot codepath and can get pretty big, so instead of printing to STDOUT, it will write to the custom C</var/log/christmas/reindeer.data> file.
```

Comment:
```suggestion
Finally, calls to C<p()> made inside C<Reindeer.pm> would not attempt to stringify objects that have a subroutine C<as_string>, C<to_string> or similar, nor will they show wrapped method modifiers or methods inherited from parent classes. Even so, that particular output is in a hot codepath and can get pretty big, so instead of printing to STDOUT, it will write to the custom C</var/log/christmas/reindeer.data> file.
```